### PR TITLE
acrn: add acrn-kernel pkg

### PR DESCRIPTION
**Pull request by jren1**
Signed-off-by: Jack Ren <jack.ren@intel.com>

### CODE REVIEW

**Comment by jren1** [comment]
Actually I copied the Docker file from pkg/kernel/Dockerfile and just made small changes. I can't understand why Yetus is failing, since those problematic Docker files lines are exactly same as pkg/kernel/Dockerfile. Any hints about how to correct those errors?

From Yetus:
pkg/acrn-kernel/Dockerfile:6 DL3019 Use the `--no-cache` switch to avoid the need to use `--update` and remove `/var/cache/apk/*` when done installing packages
pkg/acrn-kernel/Dockerfile:41 SC2046 Quote this to prevent word splitting.
pkg/acrn-kernel/Dockerfile:41 SC2039 In POSIX sh, == in place of = is undefined.
pkg/acrn-kernel/Dockerfile:41 SC2015 Note that A && B || C is not if-then-else. C may run when A is true.
pkg/acrn-kernel/Dockerfile:41 DL3019 Use the `--no-cache` switch to avoid the need to use `--update` and remove `/var/cache/apk/*` when done installing packages
pkg/acrn-kernel/Dockerfile:67 SC2046 Quote this to prevent word splitting.
pkg/acrn-kernel/Dockerfile:67 SC2086 Double quote to prevent globbing and word splitting.
pkg/acrn-kernel/Dockerfile:67 DL3003 Use WORKDIR to switch to a directory
pkg/acrn-kernel/Dockerfile:75 DL3003 Use WORKDIR to switch to a directory
pkg/acrn-kernel/Dockerfile:80 SC2046 Quote this to prevent word splitting.
pkg/acrn-kernel/Dockerfile:80 SC2086 Double quote to prevent globbing and word splitting.
pkg/acrn-kernel/Dockerfile:80 DL3003 Use WORKDIR to switch to a directory
pkg/acrn-kernel/Dockerfile:80 DL4006 Set the SHELL option -o pipefail before RUN with a pipe in it
pkg/acrn-kernel/Dockerfile:93 SC2059 Don't use variables in the printf format string. Use printf "..%s.." "$foo".

**Comment by deitch** [comment]
> I can't understand why Yetus is failing, since those problematic Docker files lines are exactly same as pkg/kernel/Dockerfile.

Yes. yetus is not set up to review the _entire_ code base, just the diffs. So even though you are just copying "something else", that "something else" isn't getting reviewed by yetus, while your "new thing" is. If we had yetus review the entire code base we would find an unmanageable number of errors, which would hold everything up for a long time, so instead it is being done incrementally. 

It can be a bit... surprising. :-)

> Any hints about how to correct those errors?

Sure. See below. Will address some of them. Each entry is the line of the Dockerfile, e.g. `pkg/acrn-kernel/Dockerfile:6` means the "offending" text is on line 6 of the `Dockerfile`.

> pkg/acrn-kernel/Dockerfile:6 DL3019 Use the `--no-cache` switch to avoid the need to use `--update` and remove `/var/cache/apk/*` when done installing packages

Just what it says, to add `--no-cache`

> pkg/acrn-kernel/Dockerfile:41 SC2046 Quote this to prevent word splitting.

It wants `$(uname -m)` in quotes as `"$(uname -m)"`; same for `x86_64` as `"x86_64"`

> pkg/acrn-kernel/Dockerfile:41 SC2039 In POSIX sh, == in place of = is undefined.

In posix sh, `[ "$(uname -m)" == "x86_64"  ]` isn't valid. It should be a single `=` as `[ "$(uname -m)" = "x86_64" ]`

> pkg/acrn-kernel/Dockerfile:41 SC2015 Note that A && B || C is not if-then-else. C may run when A is true.

You have `[ $(uname -m) == x86_64 ] && apk add libunwind-dev=1.2.1-r3 || true`, which is `A && B || C`, which is confusing and may not do what you expect. It wants you to clarify it so it is consistent.

> pkg/acrn-kernel/Dockerfile:41 DL3019 Use the `--no-cache` switch to avoid the need to use `--update` and remove `/var/cache/apk/*` when done installing packages

Same for the `apk add` on this line.

> pkg/acrn-kernel/Dockerfile:67 SC2046 Quote this to prevent word splitting.

Not 100% positive, but I believe it is complaining about `/tmp/kernel-modules/lib/modules/$DVER`, which, if `$DVER` has spaces in it, can be expanded and lead to unpredictable behaviour, so it wants you to double-quote it `"/tmp/kernel-modules/lib/modules/$DVER"`.

It also, I believe, is complaining about `DVER=$(basename $(find /tmp/kernel-modules/lib/modules/ -mindepth 1 -maxdepth 1))`, which, if the result were to have whitespace in it, could lead to strange behaviour. Just make it happy by double-quoting.

> pkg/acrn-kernel/Dockerfile:67 DL3003 Use WORKDIR to switch to a directory

It doesn't like the fact that you did `cd /tmp/kernel-modules/lib/modules/$DVER` (or `cd <anything>`), and wants you to _always_ use the Dockerfile command `WORKDIR` to change directories. I don't agree with it in this circumstances - sometimes it is perfectly fine to have multiple `cd` in a single `RUN` command - so I would leave this one for now until that is the only problem left; then we can add an override.

> pkg/acrn-kernel/Dockerfile:75 DL3003 Use WORKDIR to switch to a directory

same as previous

> pkg/acrn-kernel/Dockerfile:80 SC2046 Quote this to prevent word splitting.

Same as earlier.

> pkg/acrn-kernel/Dockerfile:80 SC2086 Double quote to prevent globbing and word splitting.

Same as earlier

> pkg/acrn-kernel/Dockerfile:80 DL3003 Use WORKDIR to switch to a directory

Same as earlier.

> pkg/acrn-kernel/Dockerfile:80 DL4006 Set the SHELL option -o pipefail before RUN with a pipe in it

It doesn't like that you used a pipe in a sh command in `RUN`, i.e. `tar cf - -T - | (cd $dir; tar xf -) ` and `'*.sh' \) | \`. It wants you to explicitly require the `pipefail` option, so that if the command before the pipe fails, the whole thing will fail. This one makes good sense. You can add `set -o pipefail` at the beginning of the run.

> pkg/acrn-kernel/Dockerfile:93 SC2059 Don't use variables in the printf format string. Use printf "..%s.." "$foo".

It doesn't like that you did `printf "KERNEL_SOURCE=${KERNEL_SOURCE}\n"`, since `printf` supports formats. It wants you to do `printf "KERNEL_SOURCE=%s\n" "${KERNEL_SOURCE}"`. I think it is being a stickler here, but it doesn't hurt to make it happy.

File: pkg/acrn-kernel/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,99 @@
+FROM alpine:3.10 as kernel-build
+#linuxkit/alpine:f2f4db272c910d136380781a97e475013fabda8b AS kernel-build
+
+RUN apk update
+
+RUN apk add \
```
Comment (rvs) [inline pkg/acrn-kernel/Dockerfile]: can you please add --no-cache here and remove the RUN apk update line?

File: pkg/acrn-kernel/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    zlib-dev=1.2.11-r1 \
+    openssl=1.1.1d-r0 \
+    lz4=1.9.1-r0 \
+    elfutils-libelf=0.168-r2 \
+    elfutils=0.168-r2 \
+    elfutils-dev=0.168-r2
+
+
+# libunwind-dev pkg is missed from arm64 now, below statement will be removed if the pkg is available.
+RUN [ $(uname -m) == x86_64 ] && apk add libunwind-dev=1.2.1-r3 || true
```
Comment (rvs) [inline pkg/acrn-kernel/Dockerfile]: This needs to be remove -- its not needed for ACRN (which will also take care of yetus complaints for this line)

File: pkg/acrn-kernel/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+# build acrn-kernel
+RUN mkdir /out
+RUN cp kernel_config_uefi_sos .config && \
+    make olddefconfig && \
+    make -j4 && \
+    cp arch/x86_64/boot/bzImage /out/kernel; \
+    cp System.map /out
+
+RUN make INSTALL_MOD_PATH=/tmp/kernel-modules modules_install && \
```
Comment (rvs) [inline pkg/acrn-kernel/Dockerfile]: Do we really need to install modules for ACRN dom0 kernel? If not -- lets remove this.

File: pkg/acrn-kernel/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    ( DVER=$(basename $(find /tmp/kernel-modules/lib/modules/ -mindepth 1 -maxdepth 1)) && \
+      cd /tmp/kernel-modules/lib/modules/$DVER && \
+      rm build source && \
+      ln -s /usr/src/linux-headers-$DVER build ) && \
+    ( cd /tmp/kernel-modules && tar cf /out/kernel.tar lib )
+
+# Headers (userspace API)
+RUN mkdir -p /tmp/kernel-headers/usr && \
```
Comment (rvs) [inline pkg/acrn-kernel/Dockerfile]: Do we really need to install modules for ACRN dom0 kernel? If not -- lets remove this.

File: pkg/acrn-kernel/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    make INSTALL_HDR_PATH=/tmp/kernel-headers/usr headers_install && \
+    ( cd /tmp/kernel-headers && tar cf /out/kernel-headers.tar usr )
+
+# Headers (kernel development)
+RUN DVER=$(basename $(find /tmp/kernel-modules/lib/modules/ -mindepth 1 -maxdepth 1)) && \
```
Comment (rvs) [inline pkg/acrn-kernel/Dockerfile]: Do we really need to install headers for ACRN dom0 kernel? If not -- lets remove this.

File: pkg/acrn-kernel/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    cp /acrn-kernel/Module.symvers $dir && \
+    find . -path './include/*' -prune -o \
+           -path './arch/*/include' -prune -o \
+           -path './scripts/*' -prune -o \
+           -type f \( -name 'Makefile*' -o -name 'Kconfig*' -o -name 'Kbuild*' -o \
+                      -name '*.lds' -o -name '*.pl' -o -name '*.sh' \) | \
+         tar cf - -T - | (cd $dir; tar xf -) && \
+    ( cd /tmp && tar cf /out/kernel-dev.tar usr/src )
+
+RUN printf "KERNEL_SOURCE=${KERNEL_SOURCE}\n" > /out/kernel-source-info
```
Comment (rvs) [inline pkg/acrn-kernel/Dockerfile]: Don't use variables in the printf format string. Use printf "..%s.." "$foo".

File: pkg/acrn-kernel/Dockerfile (hunk 1 above)
Comment (jren1) [reply pkg/acrn-kernel/Dockerfile]: @rvs agree. I will hold on this patch and  let's discuss about why there are so many patches in the acrn-kernel and how we can handle those big diffs and what our kernel upstreaming plan is.

**Comment by rvs** [comment]
Closing this since #315 got merged.

@jren1 -- please let me know if there's anything left in this PR that needs to be salvaged.

**Pull request closed without merging**